### DISCUSSION TITLE
Tropical Cyclone Image Artifact Removal/Smoothing

### DISCUSSION THREAD

**Discussion by haicyu** [CLIMADA-project/climada_python#924] [Ideas]
**Is your feature request related to a problem? Please describe.**
Spurious image artifacts are making the hazard intensity profiles for tropical cyclones unusable at a high resolution.
Tropical cyclone storm tracks are printed into hazard intensity profiles at high spatial resolution with low temporal resolution. This results in spatial artifacts instead of a smooth profile of a tropical cyclone traveling through space over time. Is there any way to plot a smoother intensity profile for a given tropical cyclone?

**Describe the solution you'd like**

**Describe alternatives you've considered**
I've increased the number of synthetic tracks generated to get more smoothness by having them overlap with each other, but I've reached the limits of my system. Currently my model takes 4 hours to build, and I am still seeing these image artifacts. Is there any other way to get some smoothing? We should get a smooth intensity profile for a tropical cyclone along the direction of tropical cyclone travel and in a general region.

**To Reproduce**
Steps to reproduce the behavior/error:
1. Run CLIMADA at high resolution, with spacing between centroids less than 0.1 degrees.
2. View TropicalCyclone intensity map.

**Additional context**

**Screenshots**

![Screenshot 2024-07-15 135103](https://github.com/user-attachments/assets/7658d4e1-8078-443e-8cc6-b8d3a32367a8)
![Screenshot 2024-07-15 135039](https://github.com/user-attachments/assets/156e0ba4-4d14-430a-8129-950081173206)
![Screenshot 2024-07-15 135127](https://github.com/user-attachments/assets/a262d454-d74b-4782-8806-2640b9164150)

**Additional context**
From the view of Tokyo, we can see the expected versus actual intensity profiles along the direction of travel of the tropical cyclone, which is the red line. Because we seem to have large time steps between cyclone footprints, we have spatial artifacts in the form of these dips instead of a smooth profile. Is there any way to get continuity here?

When we look at the larger view of Japan, we have an area with lots of artifacts within a high intensity area. Is there any way to smooth out that whole region to be roughly even intensity, since the dips in intensity are accidental and not essential.

![Screenshot 2024-07-15 135039](https://github.com/user-attachments/assets/b79be313-efc7-4d57-82ef-50a1ab7c232c)
![Screenshot 2024-07-15 135127](https://github.com/user-attachments/assets/842700c9-00fe-4ca9-b909-6412e9369503)
![Screenshot 2024-07-15 135103](https://github.com/user-attachments/assets/7944af88-6b53-4d9e-87a4-243357d7e1eb)

**Comment by peanutfun** [comment]
From #916:

> I can give a few of them a try and if anything useful comes out of it, what would be a good way to share back with you?

@haicyu It would be great if you could share your results here. I suggest we compare the non-smoothed to the smoothed hazard intensity, for a single track and varying time steps, visually. Would that be possible?

**Reply by haicyu** [reply]
Sure, I am currently working on another project in CLIMADA, related to River Flooding, but when I get free I will come back to this. 

In the mean time I wonder if I could trouble you with a few questions from my initial investigation:
- Is there a quick way in CLIMADA to generate a 2D np array with the intensity values rasterized? Currently it seems to me that I would have to manually look up lat, lon, and intensity by index, form tuples, and then manually assign them to a 2D numpy array that represents the image we see when we use the method .plot_intensity().
- Could I use the first row of the .intensity field as the windspeed, instead of manually computing them from save windfields? Do you happen to know if they are the same? It seems to me that the plot_intensity() method is using data from the .intensity field, but I could be mistaken. The .intensity field also has five rows of length len(centroids), it seems that row [0] is corresponding to the image that we see based on a few experiments that I ran - I couldn't find any documentation on the other rows so I'm just curious if you happened to know what they represented.

**Comment by peanutfun** [comment]
From #916:

> Do you have any high level considerations that would be helpful to take into account?

@haicyu For me, this relates mostly to ease of use and implementation. As I said, would be happy to incorporate a suitable smoothing algorithm. However, Climada has been stable for quite some time and the lack of such a tool was never a major issue. I would therefore make its application optional, e.g. via calling an additional method after constructing the wind field, or via an additional parameter to a current method.

I assume such a smoothing algorithm requires the full wind field at every time step and position. These are computed within the `compute_windfields_sparse` function, see https://github.com/CLIMADA-project/climada_python/blob/46c48524fe69283dfe44a5259594dc6ffbee8f58/climada/hazard/trop_cyclone/trop_cyclone_windfields.py#L992
The full wind field data is then discarded unless the user chooses to `store_windfields`. In any case, the data is then only available as a sparse matrix. What type of input data do your algorithms need? Do we need to convert the data into a dense numpy array or even an xarray DataArray?

**Reply by haicyu** [reply]
A dense numpy array would be required as an input. I'm believe that the windspeeds are a single scalar value at each location that reflect the intensity of the wind, and we don't have any directional information to take into consideration?

Yes, agree. It would be easiest for development and testing purposes to keep it as a separate step, optional for the end user. I am aware of some need for smoothing, also in my own use case, and I suppose that it would be more commonly desired. Are there any specific use cases you might have in mind for such a smoothing algorithm?

**Reply by peanutfun** [reply]
There should be directional information. In the line of code I mentioned above, the wind fields have two values at each position, I guess these are the longitudinal and latitudinal wind speed vector components. The norm over the last axis then yields the magnitude of the wind speed.

> Are there any specific use cases you might have in mind for such a smoothing algorithm?

Not specific ones, but removing the nonphysical, bow-shaped artifacts would be reason enough for me to look into this further. However, I am no expert on tropical cyclone (impact) modeling. I'll try to get input from others in our team working with TCs.

**Reply by chahank** [reply]
I am not sure to understand this conversation. For tropical cyclones, you just have to make the interpolation time sufficiently short `TropCyclone.equal_timestep()`. I am not sure what would be gained by having another extra smoothing algorithm (this is already exactly what the time interpolation does, simply on the track level, taking into account direction and wind speed). 

This is also clearly explained in the tutorials.

**Reply by peanutfun** [reply]
@chahank Computing the wind fields for short time steps can be very costly. A smoothing algorithm might remove the bow-like shapes in the intensity footprint much more efficiently than choosing a shot time step.

**Reply by chahank** [reply]
While I see the argument in principle, it remains to be seen whether this is true. Smoothing is not so simple, in particular, if one wants to keep physics ok. I am not against it, but I have high doubts that it is worth the time (since there is already a working solution), and I cannot see what one you could publish on that. But one is always free to prove me wrong!

**Reply by peanutfun** [reply]
@chahank Please see @haicyu's comment: https://github.com/CLIMADA-project/climada_python/issues/916#issuecomment-2238416071

**Reply by chahank** [reply]
I have seen this. I am still confused. But I am happy to see a proposal!

**Reply by haicyu** [reply]
You are right @chahank that increasing time resolution can help to create smoothness in the direction of the TC's travel and thank you for your additional perspective, however as @peanutfun points out it is costly. But then from a practical perspective there remains the question of the gaps that emerge in between two parallel TC tracks, as you can see in the pictures that I have attached. 

The dips in intensity, while representing a physical simulation, do not represent the climate risk that is being faced at each location. From an industry perspective, this makes it unusable. I am aware that suboptimal technical approaches with consequential tradeoffs are being taken in order to achieve a degree of smoothness.

While the workaround may be to increase the number of synthetic tracks generated, this is proving to be too costly to be effective. Focusing on just one sub-region of the world and at the limits of my computational resources, I was able to generate the attached images which still suffer from these artifacts which make it not suitable for our use case.

Smoothing is computationally far cheaper to produce the same result - by smoothing a few tracks, you could have the same effect as having simulated hundreds of tracks that still may not adequately fill those inter-track gaps, and it allows you to operate at a lower time step resolution while achieving the same result.

But I share your concern for the physicality of smoothing, particularly at the land-sea boundary, and I would be keen to learn additional concerns that you may have and brainstorm ways to address it. I will likely do this smoothing work in the coming weeks, certainly adequate from a risk management perspective and a significant improvement on the local state of the art, but I thought it might be nice to share it with others so we all have it, it feels like a nice thing to do.

**Reply by haicyu** [reply]
Oh I just saw your final comment. Sure, happy to put something together - it may be in a few weeks time but it may be more clear when we are talking in more detail. Wish you all the best in the meantime!

**Reply by chahank** [reply]
> But then from a practical perspective there remains the question of the gaps that emerge in between two parallel TC tracks, as you can see in the pictures that I have attached.

> The dips in intensity, while representing a physical simulation, do not represent the climate risk that is being faced at each location. From an industry perspective, this makes it unusable. I am aware that suboptimal technical approaches with consequential tradeoffs are being taken in order to achieve a degree of smoothness.

This is an interesting point to me, but I am a bit confused as to what the smoothing then could be used for. If you want to assess risk using a probabilistic event-based approach, you cannot just smooth out tracks, as this will destroy your event statistical analysis such as the average annual risk or the 1/100 year risk (besides being unphysical). What kind of results are you interested in that would benefit from a smoothing?

**Reply by chahank** [reply]
Maybe a more practical approach would be to compute a large number of tracks at lower resolution, and then use an up-scaling algorithm.

**Reply by peanutfun** [reply]
>  But then from a practical perspective there remains the question of the gaps that emerge in between two parallel TC tracks, as you can see in the pictures that I have attached.

@haicyu It seems I misunderstood your initial feature request. I thought you meant the gaps in the wind field of a *single track*, that emerge when time steps are so large that the cyclone travels several centroids within each time step. I think that these can indeed be smoothed out by an algorithm and actually improve physical consistency to some degree.

I did not realize you meant the overall coverage of the synthetic tracks. Note that generating a climatological, probabilistic TC track set is a major scientific undertaking and there are several approaches to it. The one implemented in CLIMADA is likely not competitive to the state of the art. Depending on your specific use case, it might be better to look into existing TC hazard datasets like e.g. STORM https://doi.org/10.1038/s41597-020-0381-2

**Comment by haicyu** [comment]
@peanutfun would you on the CLIMADA side have any resources available for a review? Would this be something that could be published in a paper?

**Reply by peanutfun** [reply]
What exactly do you mean by resources? We should have enough time to spare for a thorough code review. This usually happens within a GitHub pull request, see our [contribution guide](https://github.com/CLIMADA-project/climada_python/blob/main/CONTRIBUTING.md).

In the past, we have published submodules of Climada in journals like Geoscientific Model Development (see https://doi.org/10.5194/gmd-15-7177-2022) and we have a calibration submodule currently under review in the Journal of Open Source Software (see https://joss.theoj.org/papers/50845e31c6cb6894baae492bfd853671). The latter might be a suitable place to publish the entire TC submodule with the different wind field models implemented. However, we would have to start a conversation with the original module authors about that.

**Comment by bguillod** [comment]
To try and summarize, there are three sources of lack of smoothness in your example @haicyu, which I'm very familiar with (most of those have already been mentioned by @peanutfun or @chahank in some flavor; I'm just trying to summarize as an overview here):

1. **Along a single track**: If the temporal resolution of the track was relatively low before calculating the wind map. I personally typically use tracks interpolated to 10 minutes or lower (`time_res_h=1/6`).
2. **Between each set of two tracks**: this can be partly solved by generating more synthetic tracks as you suggested, however there is a limit to what this can do, especially if the spatial resolution of your centroids is very high (which by the way becomes fake precision to me when you go higher resolution than about 5km or so, because then one would need to include additional information in the wind speed modelling such as roughness length and topography - note that the "5km" mentioned here come from my intuition only).
3. However, the **main issue** in your example, IMHO, is that you are looking at the maximum intensity of the event set. This of course is very unstable and sensitive to individual tracks. Instead I recommend you to look at return period values instead, which are less sensitive - provided you have enough data (total number of simulated years) to robustly estimate the desired return values. For instance you could generate 1000 years of data (25 ensemble members with 40 historical years or something like this) and then estimate return values for return periods of 10, 20 or 50 years rather robustly.

I agree with @chahank that smoothing the values via an algorithm can be quite misleading in this case so I'd be cautious with such procedures.

**Reply by haicyu** [reply]
Thank you for your input @bguillod @chahank @peanutfun 
Yes, @bguillod you provide a good summary of the two sources of the lack of smoothness.
When it comes to risk management, if there is a significant difference between the hazard faced by two adjacent properties, we should have a good reason why, so in general we are expecting smoother maps.
I am less familiar with the return period method you mention in #3. Do you have any worked examples that show the result? I am wondering if this approach would meet the needs of my use case, and if there is anything like a tutorial that I could follow.
Although the maximum intensity output results generated by CLIMADA may be unstable, the industry standard for risk management of TC hazards is to look at the maximum intensity that a property could face, and calculate the damages that could result as the first step. Windspeed intensity is the most relevant metric here.
@chahank could you elaborate on what you mean by up scaling algorithm? I am open to exploring this as an option as well.
@peanutfun we are really impressed with CLIMADA's open source capabilities and are happy to stay with the current algorithm. We seem to be very close to making it work great with our use case, and we are already using it in some form, so we will like to stick with what CLIMADA makes available. Which is also why we are interested in offering an improvement to CLIMADA if it is relevant to you and desireable.

**Reply by haicyu** [reply]
![Intensity_Tokyo_2022_RCP26_emanuel-zoom_1-0 05_n_synth10](https://github.com/user-attachments/assets/3a8fddf3-92b7-4eb9-ad38-a08d0abf1142)
I used time_step_h=1/6 and I got the above image. It produces smoothing to address issue 1 as mentioned, but there remain some 'ribcage' artifacts as well as the narrow inter-track dips in intensity which we would like to get rid of. This was a helpful step forward though.

**Reply by haicyu** [reply]
It looks like some ring artifacts still cannot be avoided with a higher time resolution, this for Fuzhou, China:
![Intensity_Fuzhou_historical_RCPunweighted_emanuel-zoom_1-0 05_n_synth10](https://github.com/user-attachments/assets/20f2db89-72e7-4bc9-ba3d-10d5aeeabed9)

**Reply by bguillod** [reply]
To calculate the local return value for a given return period you can use the `Hazard`'s method `local_exceedance_inten`, or the `plot_rp_intensity` method to directly plot it.
If you e.g. select a return period of 100 (years), you get an estimate of the wind speed which would occur with a probability of 1/100 every year. This is much more robust than the maximum of all the values; moreover you can select to look at occasional events (e.g. for a return period of 10 or 20 years), rare events (e.g. 50 years) or very rare ones (e.g. 100 years or higher) (n.b. subjectivity is involved in the choice of the appropriate return period, but this is still much better than to rely on the maximum, since a maximum could be something happening every million years on average, which is probably overkill unless you're tackling something like nuclear power plant safety; also, the maximum is always the maximum _of the data subset you are using_ and is likely to display a lower value if this data consists of 20 years compared to if it consists of 20'000 stochastically simulated years).

**Reply by haicyu** [reply]
Hi @bguillod,

I tried this method as you suggested and the results look good. Can these return period intensity maps be used with impact functions to calculate AAL?

![no-smooth](https://github.com/user-attachments/assets/a43a5cb9-04d2-4d60-9dda-b43a26f3645e)
![rp-plots](https://github.com/user-attachments/assets/76801c93-d2ac-444f-bd46-515ef2abe3d9)

Thanks,
Charles

**Reply by bguillod** [reply]
> I tried this method as you suggested and the results look good.

Great!

> Can these return period intensity maps be used with impact functions to calculate AAL?

In theory yes, you could approximate the AAL using return period maps. However, in this case it seems a convoluted way to do it: the most natural but also straightforward way to calculate the average annual loss (AAL) is to calculate the impact from your hazard data (`TropCyclone`) and then extract the AAL (in climada, the AAL at each exposure location is the attribute `eai_exp` from an `Impact` object, while the AAL of the whole portfolio is attribute `aai_agg`).

**Reply by haicyu** [reply]
> to calculate the impact from your hazard data

Yes, this is what I have done but it appears to be quite sensitive to location. When I set a portfolio with a single exposure and move it around where there is a storm footprint, I get a high rate of change of AAL with respect to space, hence what brought me here with the need for smoothing. Is there any way to compute a "smoothed" AAL? That would be the end goal for me.

**Reply by bguillod** [reply]
> When I set a portfolio with a single exposure and move it around where there is a storm footprint, I get a high rate of change of AAL with respect to space

Is this when using a full probabilistic event set with many years, or are you just using a few events for this? I would expect that with a large enough event set (i.e. one which is built with many years of historical tracks **and** enough synthetic tracks per historical track **and** all tracks being interpolated to a sufficient temporal resolution before calculating the wind fields) the AAL should be relatively smooth.

It is difficult to judge without looking into your data in detail, however such an investigation seems out of scope of this github issue, unless it relates to an issue in the climada code.
